Add resource allocator api

mrjana opened the pull request:
Added resource allocator api to help worker nodes to attach and detach
to a swarm created network so that individual containers can be
independently attached to a swarm network and discover/access swarm
services directly.

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

aaronlehmann commented on manager/manager.go:
```diff
@@ -308,6 +311,7 @@ func (m *Manager) Run(parent context.Context) error {
 	api.RegisterRaftMembershipServer(m.server, proxyRaftMembershipAPI)
 	api.RegisterControlServer(m.localserver, localProxyControlAPI)
 	api.RegisterControlServer(m.server, authenticatedControlAPI)
+	api.RegisterResourceAllocatorServer(m.server, authenticatedResourceAPI)
```
No raft proxy for this service? Will clients be expected to connect directly to the leader?

mrjana replied:
No need for raft proxy. The only clients are the agent sessions and these apis require node information to find out which node this request is coming from. So these requests can't originate from anywhere.

aaronlehmann replied:
I think you do need the proxy. The agent doesn't know which manager is the leader. When it connects to the dispatcher, it connects to any manager, and relies on its raft proxy to redirect the RPCs. I think the same would apply for `ResourceAllocator`.

The node information is forwarded along with forwarded RPCs. `RemoteNode` should work transparently.

mrjana replied:
Hmm.. May be I misunderstanding something. Doesn't the follower manager which forwards the requests do it automatically? Is anything need to be setup? I thought the only other thing needed is a localProxy for un-authenticated unix socket listener for direct communication with the leader.

aaronlehmann replied:
It is automatic if you wrap the service with `NewRaftProxyResourceAllocatorServer`. Notice that the dispatcher has:

```go
proxyDispatcherAPI := api.NewRaftProxyDispatcherServer(authenticatedDispatcherAPI, cs, m.RaftNode, ca.WithMetadataForwardTLSInfo)
```

I would follow the dispatcher example since this service is similar.

Note that the proxy should wrap the authenticator, rather then vice versa. This is so that authorization always happens on the leader.

mrjana replied:
Ok. Will fix it.

aaronlehmann commented on api/specs.proto:
```diff
@@ -77,14 +77,14 @@ message ServiceSpec {
 		repeated string aliases = 2;
 	}
-	repeated NetworkAttachmentConfig networks = 7;
+	repeated DeprecatedNetworkAttachmentConfig deprecated_networks = 7;
```
Perhaps this should use `[deprecated=true]` rather than (or in addition to) renaming the type to `DeprecatedNetworkAttachmentConfig`.

mrjana replied:
Yeah I wasn't sure which way to go so I threw it in here like this and get opinions.

stevvooe replied:
No need. See my comment above.

mrjana replied:
Ok, will retain the old name as is.

stevvooe replied:
So, this field will be:

```proto
repeated NetworkAttachmentConfig networks [deprecated=true];
```

aaronlehmann commented on manager/resourceapi/allocator.go:
```diff
@@ -0,0 +1,123 @@
+	}
+
+	if !network.Spec.AllowContainers {
+		return nil, grpc.Errorf(codes.PermissionDenied, "cannot attach to network %s(permission denied)", request.Config.Target)
```
There should be a space between `%s` and `(permission denied)`.

mrjana replied:
Will fix.

mrjana commented on manager/resourceapi/allocator.go:
```diff
@@ -0,0 +1,123 @@
+		}
+	})
+	if network == nil {
+		return nil, grpc.Errorf(codes.NotFound, "network  %s not found", request.Config.Target)
```
Thanks. Will fix it.

mrjana commented:
docker integration changes for this functionality is here https://github.com/mrjana/docker/commit/3ccf2e605b76c2921c38cbe9964c8dccb67417f0 for reference.

stevvooe commented on api/specs.proto:
```diff
@@ -77,14 +77,14 @@ message ServiceSpec {
 	// instructing Swarm on how this service should work on the particular
 	// network.
-	message NetworkAttachmentConfig {
+	message DeprecatedNetworkAttachmentConfig {
```
No need to deprecate this as long as the field types and numbers match up, which they do below.

stevvooe commented on api/types.proto:
```diff
@@ -208,14 +208,14 @@ message Mount {
-		// 	"nr_inodes": true,
-		// 	"nr_blocks": true,
-		// 	"mpol":      true,
+		//	"":          true,
```
Turn off the formatting.

mrjana replied:
Argh. Formatting. Will remove.

stevvooe commented on agent/agent.go:
```diff
@@ -298,6 +298,44 @@ func (a *Agent) withSession(ctx context.Context, fn func(session *session) error
 	}
 }
 
+// AttachNetwork sends a container attach network request towards a
```
Wouldn't this be done as part of the backend? The agent shouldn't really provide this API.

mrjana replied:
We need the agent session because the server needs the node info for both authorization purposes as well as to get the node info where this request originated from (so that the task can be bound to that node). We don't need have that method here but then we need to export `withSession` so that the backend can use it.

stevvooe replied:
@mrjana The main issue is that we are coupling network attachment to the agent type: https://github.com/mrjana/docker/commit/db39e4d435e28a3cca67f124ab906cd413f6d207#diff-9cb7240401da846ac7c2252759538e9aR1213.

Let's have the agent return a resource allocator interface that can be passed to the particular component. I want to be able to change the agent exported functions without having to change daemon code.

mrjana replied:
👍 Will move it to a new interface and return that interface to backend.

stevvooe commented on agent/agent.go:
```diff
@@ -298,6 +298,44 @@ func (a *Agent) withSession(ctx context.Context, fn func(session *session) error
 
+// AttachNetwork sends a container attach network request towards a
+// manager through the already existing session.
+func (a *Agent) AttachNetwork(ctx context.Context, target string, cid string, addresses []string) (string, error) {
```
It might be a good idea to document these parameters.

mrjana replied:
Sure.

dongluochen commented on api/specs.proto:
```diff
@@ -103,6 +103,7 @@ message GlobalService {
 
 message TaskSpec {
 	oneof runtime {
+		NetworkAttacherSpec attacher = 8;
```
I feel a little uncomfortable to put `NetworkAttacherSpec` as a `runtime`. Isn't it part of container runtime?

mrjana replied:
@dongluochen This is the new runtime we are creating for attaching docker unmanaged containers(created using `docker run`). In this runtime we don't actually manage the container itself but just the attachment.

aluzzardi replied:
nit: naming. This should somehow be a noun (containerspec, unikernelspec, ...)

mrjana replied:
It is a now isn't it? :-) `NetworkAttacher`

stevvooe replied:
@mrjana `NetworkAttaché`

stevvooe replied:
Seriously: `attachment` or `network_attachment` is fine here.

mrjana replied:
Ok you got it.

aluzzardi commented on api/specs.proto:
```diff
@@ -234,6 +246,14 @@ message NetworkSpec {
+	// attach to this network and access the services attached to
+	// this network. If this flag is not enabled(default case)
+	// independent containers cannot attach to this network.
+	bool allow_containers = 6;
```
nit: Naming. I'm sorry I don't have a good suggestion, but this one is important since I assume will be visible from engine API and CLI flags.

"allow containers" seems to imply that without that, containers cannot attach to the network (even for services).

attachable? I don't know.

/cc @diogomonica

mrjana replied:
Yes, this will be visible with the same name in engine api and cli so it is important that we all agree on the name and I am open to suggestions :-)

dhiltgen replied:
I don't have a silver bullet for the name, but here are a few ideas:

* global_network
* open_to_all_nodes
* worker_visible
* allow_worker_attach

stevvooe replied:
> global_network

The implications of this name are way too wide. Does that mean the network is globally visible?

> open_to_all_nodes

We may want to scope this in the future to a specific set of nodes.

> worker_visible

All networks are "visible" to the worker.

> allow_worker_attach

Will this feature always be limited to a worker? What if I want to have a host that can attach to these networks but not run workloads? That should be possible, without changes.

The term `Attachable` is strictly true. This is a network to which one can attach.

mrjana replied:
But we do already attach services to these networks and available without any flag. So may be `ExternallyAttachable`?

stevvooe replied:
`ManualyAttachable`.

aaronlehmann replied:
`ManuallyAttachable`

diogomonica replied:
What about `public`?

mrjana replied:
@diogomonica That would be mis-interpreted as the network being a public network. It is still a private network. 

I think I am fine with `ManuallyAttachable` but it may be a little awkward in the cli. But if there are no better options I am going to go ahead with that.

diogomonica commented on manager/resourceapi/allocator.go:
```diff
@@ -0,0 +1,123 @@
+		return nil, grpc.Errorf(codes.NotFound, "network  %s not found", request.Config.Target)
+	}
+
+	if !network.Spec.AllowContainers {
```
We need a positive and negative test for this.

mrjana replied:
Do you mean we need test code to test for both positive and negative case for this?

mrjana commented:
@aaronlehmann @stevvooe @dongluochen @aluzzardi @diogomonica Taken care of comments. PTAL.

stevvooe commented on agent/resource.go:
```diff
@@ -0,0 +1,69 @@
+	// addresses to be assigned to the attachment. AttachNetwork
+	// returns a unique ID for the attachment if successfull or an
+	// error in case of failure.
+	AttachNetwork(ctx context.Context, target, id string, addresses []string) (string, error)
```
Small nit: `id` should come first. That is the "destination".

mrjana replied:
Ok. Will change.

stevvooe commented:
LGTMish

Reaction: "I can't believe this actually works."

mrjana commented:
@stevvooe Works like a charm across all the nodes in the cluster. Discovery of the containers work and they can actually connect to the service and best of all the services in the same network gets load balanced from this container. Works with the same guarantees as tasks for node going down and network partitioning scenarios.

stevvooe commented:
@mrjana Is there a way to set aliases for the created endpoint?

mrjana commented:
@stevvooe Yes it is already supported(`net-alias` option in CLI) but it doesn't need to be handled at manager level though.

stevvooe commented:
@mrjana Does the gossip populate elsewhere?

Really need to bring networking into swarmkit.

mrjana commented:
Yeah gossip actually does the work of propagating routing and discovery state  across the cluster.

aluzzardi commented on api/resource.proto:
```diff
@@ -0,0 +1,34 @@
+
+message AttachNetworkRequest {
+	NetworkAttachmentConfig config = 1;
+	string id = 2 [(gogoproto.customname) = "ID"];
```
Is there a way to do this backwards? As in, rather than the agent sending an ID up, we let the agent use the ID that comes down (`attachment_id`).

If so, it would greatly simplify a few things (for instance, no need for `NetworkAttachmentSpec`)

mrjana replied:
I did have an original implementation like that before but with that the engine needs to remember the mapping between attachment_id and container ID which means it needs to persist with that during node restart. If we make it part of this spec, it is part of the task object which gets persisted in the agent and the current agent design of trying to reconcile with the manager on node restart works pretty well for these type of tasks well.

aluzzardi replied:
I see. It's a bit overkill since, in order to persist on the agent, we're going to persist in Raft as well and expose that into the protocol. But yeah, I don't see a better solution right now.

Unless the agent has already facilities to do "local annotations" on tasks or something? /cc @stevvooe

aluzzardi commented on manager/resourceapi/allocator.go:
```diff
@@ -0,0 +1,123 @@
+	}
+
+	if !network.Spec.ManuallyAttachable {
+		return nil, grpc.Errorf(codes.PermissionDenied, "cannot attach to network %s (permission denied)", request.Config.Target)
```
nit: "Network is not manually attachable" so it's more actionable

mrjana replied:
Ok will change.

dongluochen replied:
Also update function description to include permission denied.

mrjana replied:
Ok

aluzzardi commented:
I have a few questions that came to mind when looking at the workflow:

What happens when the node goes down?
Is the task getting re-created or is it lost?
How about garbage collection? The task reaper works on services I believe and as far as I know the only way to delete tasks is by detaching

aluzzardi commented on api/specs.proto:
```diff
@@ -84,7 +84,7 @@ message ServiceSpec {
 		repeated string aliases = 2;
 	}
-	repeated NetworkAttachmentConfig networks = 7;
+	repeated NetworkAttachmentConfig networks = 7 [deprecated=true];
```
How are we handling the deprecated field?

Should it be migrated once and for all or should we have code able to deal with both?

stevvooe replied:
> Should it be migrated once and for all or should we have code able to deal with both?

We should handle and broadcast both.

Really, the deprecation plan should be placed on the field description.

mrjana replied:
We need to migrate the state from old state to new state. I haven't added it as part of this PR but we should do it before 1.13. May be we can track it as a separate issue?

aluzzardi commented:
Also - could you simulate a netsplit and see what happens?

stevvooe commented on api/specs.proto:
```diff
@@ -84,7 +84,7 @@ message ServiceSpec {
 		// Aliases specifies a list of discoverable alternate names for the service on this Target.
 		repeated string aliases = 2;
 	}
-	repeated NetworkAttachmentConfig networks = 7;
```
@mrjana I think you misunderstood me: the inner type `NetworkAttachmentConfig` is no longer needed. `NetworkAttachmentConfig` is structurally identical.

mrjana replied:
Yeah. I think I should be ablet to remove the definition of the embedded `NetworkAttachment` but keep the `networks` field there.

stevvooe replied:
Exactly.

mrjana replied:
@stevvooe The one problem with removing the embedded `NetworkAttachmentConfig` is that the type encoding of state from old version cannot be ready any more because I think the type encoding of the non-embedded version of this type will differ from the embedded version.

mrjana commented:
@aluzzardi When the node goes down the task is not recreated or lost but just stays there. When the node comes back it will reconcile with the state of the container at that point and will destroy the task if needed. This works well for the netsplit case as well. The only problem that remains is when node goes down and never comes back up. In that case the IP is not released at all. But this is the same situation that we are in for services (see #440). We need to run a general purpose GCs on such tasks whose nodes have been down for certain amount of time and reap them.

mrjana commented:
@aluzzardi I did a netsplit simulation test already by bringing down the link and since the design is such a way that the manager doesn't do anything until it hears back from agent, in this case when I brought up the link the agent reconciled and since the container is still running did not cause any state changes in the manager.

dongluochen commented on manager/resourceapi/allocator.go:
```diff
@@ -0,0 +1,123 @@
+			return grpc.Errorf(codes.NotFound, "attachment %s not found", request.AttachmentID)
+		}
+		if t.NodeID != nodeInfo.NodeID {
+			return grpc.Errorf(codes.PermissionDenied, "attachment %s doesn't belong to this node", request.AttachmentID)
```
Why the error is `PermissionDenied`?

mrjana replied:
Because the node is not authorized to delete this attachment because it did create the attachment in the first place.

stevvooe replied:
How will orphan attachments get cleaned up?

mrjana replied:
Only the node which created the attachment will be able to remove it. In case a node which created an attachment went down and never came up we have to GC that attachment in manager after a certain amount of time since the node has gone down. Note that this is the same problem that we have for service tasks (see #440)

aluzzardi commented:
> The only problem that remains is when node goes down and never comes back up. In that case the IP is not released at all. But this is the same situation that we are in for services (see #440). We need to run a general purpose GCs on such tasks whose nodes have been down for certain amount of time and reap them.

@mrjana It's a bit different - when this happens, we set the task desired state to `SHUTDOWN`.
We could change the allocator not to consider tasks that are set to shutdown.

However, I don't know what the task state of attachment tasks is.

Also, we do have a task reaper. When a node goes down for instance:
- Mark the task as `SHUTDOWN`
- Create another task
- After repeating the above a few times (5 total tasks), the task reaper deletes old tasks

So in theory if a node goes down (even without #440), **eventually** we're going to release the tasks.

Could you try out the above use cases?

I'm wondering if the reaper could be changed to handle service-less tasks.

/cc @aaronlehmann

mrjana commented:
May be @aaronlehmann can chime in. Bit yeah I think it does make the desired state SHUTDOWN but it will not remove the task until it hears back from the agent that the shutdown is complete. If that is not the case we have a bigger problem even for service tasks.

mrjana commented:
@aluzzardi Also we already make sure that the task is truly dead (as learned from agent) before deallocating the task(https://github.com/docker/swarmkit/blob/master/manager/allocator/network.go#L473). So as long as we are not reaping the task we should be good.

aaronlehmann commented:
> Also, we do have a task reaper. When a node goes down for instance:

> Mark the task as SHUTDOWN
> Create another task
> After repeating the above a few times (5 total tasks), the task reaper deletes old tasks
> So in theory if a node goes down (even without #440), eventually we're going to release the tasks.

Currently the task reaper ignores any tasks that aren't associated with a service. So it won't act on tasks created by this resource allocator.

> I'm wondering if the reaper could be changed to handle service-less tasks.

It could, but the logic would be quite different from service-oriented tasks. I'm not sure what the correct logic would be, since there's no continuity between different tasks outside the service world.

mrjana commented:
After looking at the code that is my understanding as well. Reaper does not touch service less tasks and so the task does not even move to `SHUTDOWN` state. 

Actually the only thing that we need to fix is to fix issue #440 and we are good imo.

aluzzardi commented:
@mrjana I think #440 would just hide the symptoms. There'd still be a task, it's just orphaned (so we're leaking only the task rather than task+IP).

I think @aaronlehmann would be better suited than me to review the overall design on this part - mind taking a pass?

Basically, I'm wondering what are the consequences of service-less tasks. Specifically, desired state, what happens when a node goes down (temporarily and permanently), reaping and so on.

aaronlehmann commented:
The main question for me is whether it's important to keep the task around if the node appears to be down. If it's not, the task could get automatically deleted when the node is marked as down by the dispatcher. Either we could add something new in the orchestrator to manage serviceless tasks, or the dispatcher could do it directly.

If the task does need to be kept around long-term even though the node appears to be down, then we need some kind of reaping strategy. In this case, it's not obvious to me when the task should be cleaned up.

mrjana commented:
The problem as far as releasing the IP early is the same for service tasks or service less tasks. So unless we hear from the node that this task/container has gone down we cannot deallocate the resources. 

So the design should be to for dispatcher to mark a task `ORPHANED` if the assigned node is down long enough and that would make allocator release resources. And the reaper's job is to reap serviceless `ORPHANED` tasks immediately while for service tasks it should probably follow the task history rules.

aaronlehmann commented:
That sounds reasonable to me.

> On Aug 18, 2016, at 19:16, Jana Radhakrishnan <notifications@github.com> wrote:
> 
> The problem as far as releasing the IP early is the same for service tasks or service less tasks. So unless we hear from the node that this task/container has gone down we cannot deallocate the resources.
> 
> So the design should be to for dispatcher to mark a task ORPHANED if the assigned node is down long enough and that would make allocator release resources. And the reaper's job is to reap serviceless ORPHANED tasks immediately while for service tasks it should probably follow the task history rules.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mrjana commented:
@aaronlehmann @aluzzardi Are we looking to fix the issue of permanent node down scenario as part of this PR? Or as part of a separate PR to fix #440?

aaronlehmann commented:
I'll defer to @aluzzardi. If we don't fix it in this PR, we'd be adding another possible leak when a node goes down indefinitely. But it would be a similar class of leak to what already exists in #440, so it's not an entirely new problem.

aluzzardi commented:
We can go ahead with this as is - #440 is a 1.13/P1 so hopefully we'll get both shipped together in 1.13.

aluzzardi commented on api/specs.proto:
```diff
@@ -234,6 +236,15 @@ message NetworkSpec {
+	// access the services attached to this network. If this flag
+	// is not enabled(default case) no manual attachment to this
+	// network can happen.
+	bool manually_attachable = 6;
```
@dhiltgen Does `--manually-attachable` sounds okay to you?

dhiltgen replied:
wfm

aluzzardi commented on api/specs.proto:
```diff
@@ -72,19 +72,7 @@ message ServiceSpec {
-		repeated string aliases = 2;
-	}
-	repeated NetworkAttachmentConfig networks = 7;
+	repeated NetworkAttachmentConfig networks = 7 [deprecated=true];
```
I don't fully understand where backward compatibility is made.

Also, perhaps it would be good to have a backward compatibility unit test in `allocator_test.go` or somewhere else (as far as I can see tests got converted to the new format entirely)

mrjana replied:
The backward compatibility from API point of view will be addressed in engine-api. It will accept NetworkAttachmentConfig in both ServiceSpec and TaskSpec but will always add them to TaskSpec in controlapi.

The backward compatibility from state migration point of view is not being done as part of this PR. The design for this roughly is to do inline migration in allocator when we upgrade to 1.13 when we need to upgrade services created from old version to new version to update migrate the `NetworkAttachmentConfig` from ServiceSpec to TaskSpec. I just did not want to tackle that as well as part of this PR to reduce the scope of this PR. If it helps we can create a P1 issue to make sure we don't lose track of that.

stevvooe replied:
@mrjana This is not what we talked about yesterday. We cannot modify specs on the manager side. Aren't we doing field fallback?

aluzzardi replied:
How messy would fallback be roughly, in terms of code changes?

mrjana replied:
@stevvooe I thought you were ok with spec modification. May be I misunderstood then. 

We can do fallback but the implications are that we have to have that kind of fallback check all over the codebase. And we cannot remove them for a long time(to be fair we cannot remove the inline migration logic for a long time as well but it will be limited to allocator init time code)

mrjana replied:
In any case if we decide to tackle backward compatibility changes in a separate PR then we can agree upon the best design there.

stevvooe commented on api/specs.proto:
```diff
@@ -118,6 +107,19 @@ message TaskSpec {
+message NetworkAttachmentSpec {
+	// ID spcifies a unique ID of the entity for which this
+	// attachment is for.
+	string id = 1 [(gogoproto.customname) = "ID"];
```
After come thought, it is probably clearer to call this `ContainerID`. We should probably also include an attachment id, if that is a thing.

stevvooe replied:
If we add new types in the future, we can add id fields for them at that time.

stevvooe commented:
LGTM after clearing up the ID field names.

mrjana commented:
Updated and rebased.

aluzzardi commented:
LGTM

aaronlehmann commented:
The CI failure is unrelated. It will go away if you rebase again.

mrjana commented:
Added the fallback code and changed one of the existing test case to actually use `ServiceSpec.Networks` so we have unit test coverage for both `ServiceSpec.Networks` and `TaskSpec.Networks`

mrjana commented:
@aaronlehmann Fixed `RemoveNetwork` and `UpdateService` calls to introduce fallback logic.

aaronlehmann commented on manager/controlapi/service.go:
```diff
@@ -327,7 +327,9 @@ func (s *Server) UpdateService(ctx context.Context, request *api.UpdateServiceRe
-		if request.Spec != nil && !reflect.DeepEqual(request.Spec.Networks, service.Spec.Networks) {
+		if request.Spec != nil &&
+			(!reflect.DeepEqual(request.Spec.Task.Networks, service.Spec.Task.Networks) ||
+				!reflect.DeepEqual(request.Spec.Networks, service.Spec.Networks)) {
```
Maybe we should make an exception for transitioning from `request.Spec.Networks` to `request.Spec.Task.Networks`, if the contents of `Networks` doesn't change.

mrjana replied:
Hmm... Ok. If you want to provide users a way to migrate their services created in old model to new model. Let me rework the code to accommodate that case.

mrjana commented:
@aaronlehmann Update `UpdateService` to accommodate the case of service update migration from ServiceSpec to TaskSpec without any change in networks.

aaronlehmann commented:
LGTM
